orttraining/orttraining/core/optimizer/compute_optimizer/sceloss_flops_reduction.h

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/optimizer/graph_transformer.h"
+
+namespace onnxruntime {
+
+/**
+ * @brief Graph transformer that inserts ShrunkenGather before SoftmaxCrossEntropyLossInternal.
+ *
+ * For label sparsity, we can remove them from the compute of loss function, by inserting ShrunkenGather
+ * operators for its two inputs.
+ *
+ *   logits (float) [token_count, classes]       labels (int64), [token_count]
+ *                     \                             /
+ *                  SoftmaxCrossEntropyLossInternal(ignore_index=-100)
+ *                     /                       \
+ *   loss (shape is scalar or [token_count])      log_prob [token_count, classes]
+ *
+ * Be noted in Transformer-based models:
+ * > `token_count` usually equals with `batch size` x `sequence length`.
+ * > `classes` usually equals with `vocabulary`.
+ *
+ * Only insert ShrunkGather if all following conditions are met:
+ * 1. `SoftmaxCrossEntropyLossInternal`'s reduction MUST not be 'none', to make sure loss is a scalar.
```

Comment:
Shall we use the logic in the *.cc file that checks what the reduction is istead of what it is not?

```suggestion
 * 1. `SoftmaxCrossEntropyLossInternal`'s reduction MUST be 'mean' or 'sum', to make sure loss is a scalar.
```